### ISSUE TITLE
multiSubstitution to replace a portion of an ExprTuple

### ISSUE THREAD

**Issue by wwitzel** [PyProveIt/Prove-It#102]
In the comments of Issue #31 , I proposed the following "iterReplacement" axiom:
forall_{i, j, k in Naturals} [forall_{a_1,...,a_i, c_1,...,c_k} [forall_{b, d}
(((b_1 , ... , b_j) = (d_1 , ... , d_j)) =>
((a_1 , ... , a_i , b_1 , ... , b_j , c_1 , ... , c_k) = (a_1 , ... , a_i , d_1 , ... , d_j , c_1 , ... , c_k)))
]]

However, I'm now thinking we can use multiSubstitution.  Something like:
\forall_{n \in N+} \forall_{f, x, y | |x|=|y|=n, x=y} f(x_1, ..., x_n) = f(y_1, ..., y_n)
We had a long discussion on this last July and concluded that we could just do it as regular substitution of a tuple.  But it is not the same!  This kind of substitution should allow the substituted a portion of an ExprTuple like what iterReplacement is trying to accomplish.  But it would be done without adding an axiom and would be more powerful.  It would be utilizing operation substitution using a multi-parameter lambda expression (which probably has not been tested).  So, in principle, you could perform the substitution within different kinds of tuples simultaneously.  For example,
f : x_1, ..., x_n |-> (a + x_1 + ... + x_n) / (x_1 * ... * x_n * b)
and transform this to
(a + y_1 + ... + y_n) / (y_1 * ... * y_n * b)

**Comment by wwitzel** [comment]
There is an issue with the way that operation substitution is currently implemented that makes this multi-substitution strategy fail.  Basically, it it only set up for item-by-item substitution rather than element-by-element substitution (an iteration is an item, though it may represent many elements).  But we can revise this to work for the case where an iteration has just one iteration parameter, e.g., x_1, ..., x_n, by substituting 'x' (in this example) with the full ExprTuple of arguments for the operation substitution.  Dealing with multiple parameters that includes an iteration is not so straightforward, but we can simply avoid having to deal with that case.  After all, any lambda with multiple parameter items is equivalent to one with a single iteration parameter.  We could make the lambda's generic version be that version with a single iteration parameter (if any of the parameter items are iterations) and use that generic version when doing the substitution.  This probably wouldn't come up too often in any case, but that would cover our bases.  Also, while we are at it, we should enforce the rule that iteration parameters start with the base of the indexed variable to avoid issues/confusion in that regard.  That is, you can't use x_2, ..., x_n in the parameter list if the 'base' of 'x' is 0 or 1 (which are the typical choices).  You generally wouldn't want to do that.  The lambda is introducing new parameters -- you don't want to skip any as unbound.

**Comment by wwitzel** [comment]
I've starting thinking about these things (indexed variables, iterations, and expression tuples) in a different way and am toying with the idea of changing the approach.  I haven't settled on anything specific, but wanted to start a conversation about this new thinking.  To start with, what if an indexed variable were no different than any other operation?  That is, x_i is no different than x(i).  Then, expanding one of these out by substituting a tuple in for x would really be a matter of substituting a lambda map in for x (one that maps the index to the corresponding element).  But we'd want to have a way to translate between an ExprTuple form and a Lambda form, and we also need to determine the base (index of the first element) to do this transformation.  If 'x', say, is introduced by a universal quantifier in a form like
\forall_{x_i, ..., x_j} ...
we can do something where 
x : (some expression tuple)
transforms 'x' to a lambda map where 'i' is the first element.  For example,
x : (a, b_m, ..., b_n, c)
would get transformed to
x : {k |-> ([a | k=i]; [b_{k+m-i-1} | k \in (i+1)..(i+1+n-m)]; [c | k=j]}
assuming j=n-m+i+2.
We'd have to handle lambda maps with multiple conditions, but that would be a nice feature in any case.  This mapping may need to be split up further if, for example, the inner expression had some x_p, ..., x_q where did not match the provided partitions.

The details of this approach isn't entirely clear and may not be the best way.  Also, how would we handle something like
\forall_{x \in R^3}
rather than
\forall_{x_1, ..., x_3 \in R}
where the 'base' is not explicit in the former case?

But there are interesting possibilities here that I would like to discuss.

**Comment by wwitzel** [comment]
I still have a lot of questions about the approach we want to take.
* Should IndexedVariable be an Expression class with special features, or should this concept be handled simply as an Operation?
* Should ExprTuple implicitly represent a Lambda map -- how could we handle that conversion?
* Where do we track the 'base' regarding indexing an ExprTuple?
Let's discuss these in today's meeting.

**Comment by wwitzel** [comment]
After thinking more about these issues, may we can and should keep the derivation rules REALLY simple.  How?  Well, maybe we allow specialization like
|- \forall_{x_1, ..., x_n} f(g(x_1), ..., g(x_n))
 x : (h(1), ..., h(n))
|- f(g(h(1)), ..., g(h(n)))
but that is pretty much the extent of specialization of iterated variables that is allowed.  Internally, this could essentially be interpreted as "x : i -> h(i)".

In order to do more complicated instantiations of iterated variables, we can perform explicit transformation first.  For example,
|- \forall_{n in N} \forall_{a_1, .., a_n, b_1, .., b_n) a * b = a_1*b_1 + ... + a_n*b_n
could be used to derive
{j in N, k in N, j <= k} |- (x_1, ..., x_j, y_1, ..., y_k) * (z_1, ..., z_k, x_1, ..., x_j) 
             =  x_1*z_1 + ... + x_j*z_j+ y_{1}*z_{1+j} + ... + y_{k-j}*z_{k-j+j} 
                   + y_{1+k-j}*x_{1} + ... + y_{j+k-j}*x_{j}
through a series of transformation with via an intermediate result of
 |- \forall_{n in N} \forall_{a_1, .., a_j, a'_{j+1}, ..., a'_{k}, a"_{k+1}, ..., a"_{j+k}, 
                                      b_1, .., b_j, b'_{j+1}, ...., b'_{k} b"_{k+1}, ..., b"_{j+k}
  (a_1, .., a_j, a'_{j+1}, ..., a'_{k}, a"_{k+1}, ..., a"_{j+k)) * 
  (b_1, .., b_j, b'_{j+1}, ...., b'_{k} b"_{k+1}, ..., b"_{j+k}) 
   = a_1*b_1 + ... + a_j*b_j + a'_{j+1}*b'_{j+1} + ... + a'_{k}*b'_{k}
     + a"_{k+1}*b"_{k+1} + ... + a"_{j+k}*b"_{j+k}

Importantly, you would have full and explicit control of the indexing to get to the desired result.  Using a fancy "one-step" specialization approach, there is some ambiguity in the form of the answer.  Instead of 
             =  x_1*z_1 + ... + x_j*z_j+ y_{1}*z_{1+j} + ... + y_{k-j}*z_{k-j+j} 
                   + y_{1+k-j}*x_{1} + ... + y_{j+k-j}*x_{j}
it could have been
             =  x_1*z_1 + ... + x_j*z_j+ y_{j+1-j}*z_{j+1} + ... + y_{k-j}*z_{k} 
                   + y_{k+1-j}*x_{k+1-k} + ... + y_{j+k-j}*x_{j+k-j}
(in fact, this would probably be an intermediate in the new approach).  The point is, though, that you would take small steps without ambiguity rather than one big step which is hard to follow and ambiguous.

I'm really liking this approach.  Any objections/concerns/suggestions?

**Comment by wwitzel** [comment]
We can make the derivation rule a little more versatile (compared with the proposal above) to make proofs shorter and easier to follow with the following rule which should still be easy to implement a check.  We could allow specializing x_i, ..., x_j with x mapping to a tuple involving multiple items (singular elements as well as Iters) by

1) Checking that the Iter index ranges along with indices for the singular elements matches (i, ..., j).
    **Example 1**
        i:1, j:3, x : (a, b, c), with (1, 2, 3) = (1, ..., 3)
    **Example 2**
        i:1, j:n+2, x : (a, b_2, ..., b_n, c, d), with (1, 2, ..., n, n+1, n+2) = (1, ..., n+2)
    **Example 3**
        i:i, j:k, x : (a_i, ..., a_j, a_{j+1}, ..., a_k), with (i, ..., j, j+1, ..., k) = (i, ..., k)

2) Substitution of the inner expressions when doing this substitution works as follows:
    a) Individual indexed variables (of 'x', say) are replaced with the corresponding element of the specialization of x which must be a singular element.  For example, and in correspondence with **Example 1** above, you could first specialize i:1 and j:3 and then substitute in (x_1, x_2, x_3) for (x_1, ..., x_3).  Then with x:(a,b,c), (x_1, x_2, x_3) would become (a, b, c).  But you would not be able to do x : (f(1), ..., f(3)).  If you wanted that, you should have kept the (x_1, ..., x_3) version.
    b) Iterations involving indexed variables are replaced with the appropriate span of items of 'x' where the starting and ending indices match.  So, in the example above, we could have kept the (x_1, ..., x_3) form and specialized x:(a, b, c) to get (x_1, ..., x_3) replaced with (a, b, c).  You could also have an (x_i, ..., x_k) form and specialize x:(a_i, ..., a_j, a_{j+1}, ..., a_k) or even x:(a_i, ..., a_{j-1}, b, c_{j+1}, ..., c_k).  The splitting correctness is enforced by "1)" above; e.g., (i, ..., j-1, j, j+1, ..., k) = (i, ..., k).

This seems to provide a nice balance of flexibility and simplicity.

**Comment by wwitzel** [comment]
Deepak brought up the question of multiple indices of an indexed variable, getting ready for 2D structures.  My latest thought on this is to internally treat it as a nested structure (just like the 2D structure will be nested ExprTuples) but can display as multiple indices.  That is, internally you would have (x_i)_j, but externally it would show as x_{i, j}.

**Comment by wwitzel** [comment]
A couple of modifications w.r.t. my March 14th comment:
Let's restrict specialization of indexed variables to have the index match the parameter of any Iter that contains it, and have the start and end match its introduction in the universal quantifier.  That is,
|- \forall_{i, j \in N | i < j } \forall_{x_i, ..., x_j} x_i^2 + ... + x_j^2 >= 0.
is good to go, but 
|- \forall_{i, j \in N | i < j } \forall_{x_i, ..., x_j} x_i^2 + x_{i+1}^2 + ... + x_j^2 >= 0.
would need to be transformed before the specialization is possible.  That doesn't mean the other forms aren't allowed.  It just means that the statement must be transformed before the specialization can be done.  In the example above, we could either merge the sum or split the x_i, ..., x_j.  The former would transform (via axioms/theorems) the bottom example to the top example.  The latter would transform (via axioms/theorems) the bottom example to
|- \forall_{i, j \in N | i <= j } \forall_{a, y_{i+1}, ..., y_j} a^2 + y_{i+1}^2 + ... + y_j^2 >= 0.
via
|- \forall_{f} \forall_{i, j \in N | i <= j} 
           [(x_i, ..., x_j) |-> f(x_i, ..., x_j)] =[(a, x_{i+1}, ..., x_j) |-> f(a, x_{i+1}, ..., x_j)]
which we can derive from the proveit.logic.equality.mappingSubstitution axiom and relabeling.  Relabeling must allow for splitting iterated variables as well as other enhancements to allow relabeling inner expressions (I'll add another "issue" for that specifically).

**Comment by wwitzel** [comment]
With the simplification advocated in the previous comment, there isn't really a need for an instance variable to have an index any more complicated than a single variable (i.e., that of a Iter parameter).  This would avoid potential ambiguity: Is "x_{i+1}, ..., y_{j+1}" an Iter from "i+1" to "j+1" or "i" to "j" -- it would have to be the former with this new rule.  Going this route, you couldn't do something like
x_i*y_{i+1} + ... + x_j*y_{j+1}
But would you really need to?  If they were regular operations instead of indexed variables, there wouldn't be a limitation.  You can do
f(i)*g(i+1) + ... + f(j)*g(j+1)
no problem.  And in that mode (especially with the new Conditional feature), you can do anything you might have wanted to with fancier options for indexed variable.  Indexed variables don't really enable anything you couldn't do with general operations, they just provide convenience for specializing Iters with tuples.  But, to keep this feature simple, easy, and intuitive, let's keep the functionality somewhat restrictive unless/until we come upon a real need to enhance it.  At which point, we would be starting from something very simple and straightforward and would be motivated to make a small enhancement that maintains a lot of that simplicity.

**Comment by wwitzel** [comment]
One concern with the previous idea is that you couldn't deal with both
`z_1, ..., z_j`
and 
`z_{j+1}, ..., z_k`
iterations at the same time.  Is there something simple that can be done for that?  How do we avoid overlaps and the contradictions that may arise from that?

**Comment by wwitzel** [comment]
What we'd want to do to handle the above is to have a known truth that
{j in Naturals, k in Naturals, j <= k} |- (1, ..., k) = (1, ..., j, j+1, ..., k)
If you want to general over z which involved both z_1, ..., z_j and z_{j+1}, ..., z_k, use the above known truth to consolidate it to a single z_1, ..., z_k and make sure it covers all of the instances of z.  In fact, do this for any lambda map.  If any of its parameter variable appears as variables being indexed in the lambda body, make sure it is covered all in one.  But how would that be enforced?

**Comment by wwitzel** [comment]
We shouldn't have to worry about making sure all lambda maps play by the rules of covering all instances of indexed variables.  They are constrained to not repeat variables as parameters (including indexed variables) and that should be sufficient.  We should be okay as long as
1) We cannot prove equivalence between different lambda maps that have different meaning.
2) We cannot perform operation substitution using a lambda map operator in which the outcome is ambiguous or wrong.
But we do want to enforce the rule described above when generalizing over a variable involved in multiple iteration ranges -- make sure all of them are consolidated.  Or maybe we just make that check on the specialization/instantiation side.  The important thing is to avoid contradictions if ranges overlap.  To do that, make sure the ranges don't overlap.  To do that, make sure the ranges are all parts of a known expansion of the full coverage of iterated indices.

**Comment by deepakkapur** [comment]
I think this whole mess can be very confusing. You seem to write your comments as they come to you. Instead, there is a need to take a pause and think through issues especially 

first decide what is needed at a top level -- goal, motivation and then
second, how to facilitate it

third, how to implement it.

I would not want to do anything until the first is clear.

To start, I suggest, you read through your comments and consolidate into a single proposal which should not change for 2 weeks minimum, otherwise do not broadcast it. If you do not want to adhere to it, and have a discussion, at the very least, consolidate everything in a single proposal.

I am finding this whole process very confusing.

**Comment by wwitzel** [comment]
I'm glad you got an account, Deepak.  That will help.  I know what you are saying -- it is a bit "stream of consciousness" and sometimes I get myself confused.  I think I have it figured out and then either second-guess it or realize something I missed.  I am converging onto something and implementing it, though.  I have to implement it before I really understand it -- that's the way I operate.  Sometimes I implement something only to scrap it immediately.  That's fine, it helps me work through it.  In any case, I hope to have some demonstrations to share in the next day or two and then you can see where it is going.  Maybe we can do a call Wednesday afternoon, if I make good progress on this.

**Comment by deepakkapur** [comment]
That sounds fine.

**Comment by wwitzel** [comment]
Handled with significant-core-updates merger.
